`Development`: Improve server tests documentation (#7243)
ls1intum/Artemis: docs/dev/guidelines/server-tests.rst

************
Server Tests
************

This section covers recommended practices for writing Artemis server tests. If you want to write tests for Artemis programming exercises to test students' submissions, check out `this <https://confluence.ase.in.tum.de/display/ArTEMiS/Best+Practices+for+writing+Java+Programming+Exercise+Tests+in+Artemis>`__.

0. General testing tips
========================
Use appropriate and descriptive names for test cases so developers can easily understand what you test without looking deeper into it.

> To increase readability, prefix variable names with ``actual`` and ``expected`` instead of naming them ``int a``, ``double b``, ``String c``.

Reviewer: Maybe elaborate a bit on that since IMO there are two points here:
- prefixing variables with `actual` and `expected`
- having descriptive names

The first point is clear but maybe elaborate on the second instead of only giving examples

Author: Makes sense, I seperated it into two sentences :+1:

* Write dumb tests by avoiding the reuse of production code and focusing on comparing output values with hard-coded values.
* Invest in a testable implementation by avoiding static access, using constructor injection, and separating business logic from asynchronous execution.
* Instead of using random, use fixed test data to make error messages better understandable.
* Make use of `JUnit 5 <https://junit.org/junit5/docs/current/user-guide/#writing-tests>`__ features such as parameterized tests.
* Follow `best practices <https://www.baeldung.com/spring-tests>`__ related to spring testing.

For a more detailed overview check out `modern best testing practices <https://phauer.com/2019/modern-best-practices-testing-java/>`__.

1. Assert using the most specific overload method
==================================================

When asserting in server tests, use ``assertThat`` from the `AssertJ <https://github.com/assertj/assertj>`__ library. Another assertion statement, such as ``isEqualTo()``, **must**  follow the call. Using specific assertion statements rather than always expecting boolean values is best practice.

For example, instead of

.. code-block:: java

>     assertThat(submissionInDb.isPresent()).isTrue();
>     assertThat(submissionInDb.get().getFilePath().contains("ffile.png")).isTrue();
> 
> use the built-in assertions directly:
> 
> .. code-block:: java
> 
>     assertThat(submissionInDb).isPresent();

Reviewer 2: ```suggestion
    assertThat(submissions.size()).isEqualTo(1);
    assertThat(submissionInDb.isPresent()).isTrue();
    assertThat(submissionInDb.get().getFilePath().contains("ffile.png")).isTrue();

use the built-in assertions directly:

.. code-block:: java

    assertThat(submissions).hasSize(1);
    assertThat(submissionInDb).isPresent();
```

Author: Good idea :+1:

    assertThat(submissionInDb.get().getFilePath()).contains("ffile.png");

These assertions provide better error messages when they fail and improve the code readability. However, not all methods are suitable for this type of assertion.
If the ``isTrue`` assertion is unavoidable, specify a custom error message using the ``as`` keyword:

.. code-block:: java

    assertThat(submission.isSubmittedInTime()).as("submission was not in time").isTrue();

For more information, please read `the AssertJ documentation <https://assertj.github.io/doc/#assertj-core-assertions-guide>`__, especially the `section about avoiding incorrect usage <https://assertj.github.io/doc/#assertj-core-incorrect-usage>`__.

2. ArchUnit
===========
Use the `ArchUnit <https://www.archunit.org/>`__ library to prevent the unintentional inclusion of unnecessary packages. We use the library to enforce consistency in the code base.
Here is a simple ArchUnit test using an ArchRule to forbid JUnit assertions (in favor of AssertJ ones).

.. code-block:: java

>     void testNoJunitJupiterAssertions() {

Reviewer 2: ```suggestion
    @Test
    void testNoJunitJupiterAssertions() {
```

        ArchRule noJunitJupiterAssertions = noClasses().should().dependOnClassesThat().haveNameMatching("org.junit.jupiter.api.Assertions");

        noJunitJupiterAssertions.check(testClasses);
    }

We first define the ArchRule prohibiting the JUnit assertion package and then enforce it in test classes.

3. Counting database query calls within tests
==============================================

> It's possible to write tests checking how many database accesses a REST call performs. These tests ensure that code changes don't lead to more database calls or remind developers they do, which is especially important for commonly used endpoints.

Reviewer: ```suggestion
It's possible to write tests checking how many database accesses a REST call performs. These tests ensure that code changes don't lead to more database calls or remind developers if they do, which is especially important for commonly used endpoints.
```

************
Server Tests
************

> This section covers recommended practices for writing Artemis server tests. If you want to write tests for Artemis programming exercises to test students' submissions, check out `this <https://confluence.ase.in.tum.de/display/ArTEMiS/Best+Practices+for+writing+Java+Programming+Exercise+Tests+in+Artemis>`__.

Author: done :+1:

If the ``isTrue`` assertion is unavoidable, specify a custom error message using the ``as`` keyword:

.. code-block:: java

    assertThat(submission.isSubmittedInTime()).as("submission was not in time").isTrue();

For more information, please read `the AssertJ documentation <https://assertj.github.io/doc/#assertj-core-assertions-guide>`__, especially the `section about avoiding incorrect usage <https://assertj.github.io/doc/#assertj-core-incorrect-usage>`__.

2. ArchUnit
===========
Use the `ArchUnit <https://www.archunit.org/>`__ library to prevent the unintentional inclusion of unnecessary packages. We use the library to enforce consistency in the code base.
Here is a simple ArchUnit test using an ArchRule to forbid JUnit assertions (in favor of AssertJ ones).

.. code-block:: java

    @Test
    void testNoJunitJupiterAssertions() {
        ArchRule noJunitJupiterAssertions = noClasses().should().dependOnClassesThat().haveNameMatching("org.junit.jupiter.api.Assertions");

        noJunitJupiterAssertions.check(testClasses);
    }

> We first define the ArchRule prohibiting the JUnit assertion package and then enforce it in test classes.

Reviewer 2: Please mention the abstract super class `AbstractArchitectureTest` for ArchUnit tests here. Developers should either add their own class extending this class or add their own small general test cases to `ArchitectureTest`

Author: done

3. Counting database query calls within tests
==============================================

> It's possible to write tests checking how many database accesses a REST call performs. These tests ensure that code changes don't lead to more database calls or remind developers if they do, which is especially important for commonly used endpoints.

Reviewer 2: While often used for complete REST calls, this is not limited to rest class. We can scope this to every code block wanted by the developer.

Not sure if `unintentional performance decrements` is the best formulation here, but we should shortly explain why we want to count database requests.
```suggestion
It's possible to write tests checking how many database accesses an operation performs. These tests ensure that code changes don't lead to unintentional performance decrements and remind developers if they do, which is especially important for commonly used functionality.
```

.. code-block:: java

    class TestClass {

        @Test
        @WithMockUser(username = "instructor1", roles = "INSTRUCTOR")
        void testQueryCount() throws Exception {
            Course course = assertThatDb(() -> request.get("/api/courses/" + courses.get(0).getId() + "/for-dashboard", HttpStatus.OK, Course.class)).hasBeenCalledTimes(3);
            assertThat(course).isNotNull();
        }
    }

4. UtilServices and factories
=============================
When setting up data in tests, use helper functions from corresponding UtilService and Factory classes. We use the factory method pattern to structure test cases, depicted in the simplified class
diagram below. In this example, the ``ExerciseTestFactory`` generates exercises for tests, while the ``FileUploadTestFactory`` specializes in file upload exercises. The ``FileUploadIntegrationTest``
utilizes the services provided by ``FileUploadTestService``. The ``FileUploadTestService`` manages the creation and communication with the database.

.. figure:: resources/FileUpload_UtilService_Factory.png
    :align: center
    :alt: File upload UtilService and Factory

> In general, **UtilServices** manage the communication with the database, and **Factories** are responsible for object creation and initialization. If you cannot find the correct helper function, add a new one to the most fitting UtilService or Factory and enhance it with JavaDoc.

Reviewer 2: Can you add a small note to this section to always use the ids retuned by the database and not to assume the existence or non-existence of specific values.

Author: added a note

It's possible to write tests checking how many database accesses a REST call performs. These tests ensure that code changes don't lead to more database calls or remind developers if they do, which is especially important for commonly used endpoints.
However, we should carefully consider before adding such assertions as it makes the test more tedious to maintain.

It's possible to write tests checking how many database accesses an operation performs. These tests ensure that code changes don't inadvertently decrease performance and remind developers if they do, which is especially important for commonly used functionality.

> 

Reviewer 2: This is now basically duplicated.

Author: I now removed the previous version
